### ISSUE TITLE
[dv] More targetted FI testing for hardened FSMs and coverage

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#16572]
The current testing for hardened FSMs is standardized and just blindly injects a fault whilst some standard sequence is running.

There are a couple of issues with this:

1. The sequence may not be exploring all parts of the functionality of the block, so in particular some FSM states may simply never occur
2. On any given cycle some states may be considerably more likely than others

As a result the standardized hardened FSM tests may not stimulate state corruption in all possible states. Note that this will be not obviously picked up in coverage as the transition goes valid_state -> invalid_state -> error_state. As the 'invalid_state' is not valid it isn't part of standard automatic FSM coverage and the point of interest here involves two transitions so isn't reflected in standard automatic FSM coverage.

The current setup suffices for V2S. It demonstrates a state corruption does indeed generate the alert that it should.

For V3 we may want to be more thorough, do want to stimulate FSM state corruption in all possible states? I have a draft PR with some changes that could enable this using alert_handler for an example usage: https://github.com/lowRISC/opentitan/pull/16571

If we do want to test this we should also add coverage to ensure we've seen state corruption in all possible states.

**Comment by GregAC** [comment]
@sriyerg @cindychip @weicaiyang @msfschaffner @cdgori @vogelpi @tjaychen what are your thoughts?

**Comment by vogelpi** [comment]
Thanks for starting the discussion @GregAC , I think this is relevant and we should indeed go into that direction for V3. Your approach of waiting for a specific state to occur before injecting a fault is exactly what we do in AES. Sidenote: AES has three redundant copies of every FSM (multirail) to extend the FI hardening beyond the FSM state register to combinatorial signals and the interplay of the different controllers (e.g. cipher core FSM and main control FSM). So besides forcing the state to invalid values, we also had to force valid states in one of the FSM copies to reach the coverage threshold for V2S.

Back to non-multirail designs: Yes I think we should that. And we probably have to go beyond waiting for a particular state but also randomize the delay after entering the target state. Because sometimes the interesting stuff is happening only towards leaving the state. The question is a bit what parts can be automated and how. I can imagine that maybe coverage can be automated as part of the `PRIM_FLOP_SPARSE_FSM` macro.

**Comment by GregAC** [comment]
There's a similar issue with transitions to the error state caused by other countermeasures being triggered: https://github.com/lowRISC/opentitan/issues/16588 against we're not seeing many of these transitions due to the way we test the countermeasures.

**Comment by andreaskurth** [comment]
> And we probably have to go beyond waiting for a particular state but also randomize the delay after entering the target state. Because sometimes the interesting stuff is happening only towards leaving the state. The question is a bit what parts can be automated and how. I can imagine that maybe coverage can be automated as part of the `PRIM_FLOP_SPARSE_FSM` macro.

I think randomly picking a value from the state enum (to select the state to wait for) and then randomly waiting some cycles could make FI tests more directed to achieve sufficient coverage. I don't know if this can be generically implemented in SystemVerilog, but at least it should be generatable automatically. The randomization bounds for the number of cycles to stay in a state before injecting an error could be tricky to figure out automatically, but the wait could have `state_d != state_q` as exit condition, so that FI certainly happens during the target state.

**Comment by marnovandermaas** [comment]
I think this issue is quite important. To give a bit of context, while we looked at the FSMs in CSRNG and EDN, there were quite some intricacies in deciding what to do in illegal states and error states. For example:
- Having to go into error state when an illegal state is encountered and the module is disabled.
- Making sure that all output signals are tied off before going into the error mode.

Some extra context are these PR's:
1. EDN: https://github.com/lowRISC/opentitan/pull/16535
2. CSRNG: https://github.com/lowRISC/opentitan/pull/16397

It would be good to have automated testing to verify all of these properties.

**Comment by cindychip** [comment]
1). For standard countermeasure test, currently implementation only covers Idle -> Error transition.
     If the IP wants to cover non-idle to Error state, you can use the hook: https://github.com/lowRISC/opentitan/blob/master/hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__sec_cm_fi.svh#L15
   to implement the pre-conditions in order to hit non-idle to error states.

2). We also have FPV testbench that verifies when injecting faults to prim_sparse_fsm, prim_counter, or prim_double_lfsr, it will trigger the corresponding alerts. This should cover the transition from all states to error states.

3). But saying that, if there are any special conditions that the module wanted to cover, or other ways to reach the error state besides the countermeasures, then the fpv results are not applicable. And in that case, having specific dv or FPV testbench might be the best solution.

**Comment by sriyerg** [comment]
@cindychip and @weicaiyang can provide better context here, when we set up the sec cm testing framework, we did realize that any sparse FSM state to invalid state would be hard to cover while also having a common sec cm testing framework that works for ALL IP blocks. So we settled on the following course of action:

- Rely on the common testing framework in DV which only injects faults in the idle state and proves the alert at the IP periphery fires (and the FSM transitions to invalid state). This includes the sparse FSM (along with other standardized CMs). 
- Add the [necessary assertions](https://docs.opentitan.org/doc/project/checklist/#fpv_sec_cm_proven) in the RTL to capture ALL of the required properties  
- Add an FPV framework that is capable of injecting faults and ensure those properties do get covered (this is expected to cover all FSM states to invalid state transitions) - see [nightly results here](https://reports.opentitan.org/hw/top_earlgrey/formal/sec_cm/summary/latest/report.html)
- Exclude any FSM state to idle state transition in DV, *IFF* it is deemed sufficient that the FPV testbench sufficiently verifies the outcomes
- If there are special circuits in the design that warrants an IP block to verify in DV simulation, the transition from any FSM state to invalid state via fault injection, then we ensure that we do it on a need basis. 
- Provide hooks in the sec cm testing framwork to actuate the HW to achieve this goal - [see here](https://github.com/lowRISC/opentitan/blob/81bf4676745b84b7304a0454f2fa82b2d18cc600/hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__sec_cm_fi.svh#L14)

We have this documented to some extent here: 
https://docs.opentitan.org/doc/ug/sec_cm_dv_framework/#verification-framework-for-the-standardized-design-countermeasures

Thank you for filing this issue - it is definitely good to talk through and ensure we have all of our bases covered.

**Comment by sriyerg** [comment]
Also. see issue https://github.com/lowRISC/opentitan/issues/14407 which tracks this to some extent.

**Comment by cdgori** [comment]
I would agree that it's ok to just show proof of alert generation for V2S, but also that we need to do something more comprehensive for V3.  Checking the various non-idle -> error states and somehow measuring that each of those arcs is covered should be part of our signoff.

Historically, what I have seen is that the serious modeling/verification complexity kicks in when these FI hooks trigger simultaneously or near-simultaneously in two (or more) different modules, because the "cross product" of those error behaviors can be quite hard to scoreboard/predict.  I have seen this find some interesting/subtle bugs as well though, so the effort might be worth it.  It's also more representative of what would happen in an actual non-targeted/low-cost FI situation, because we expect a lot of simultaneous and unpredictably-clustered flips in those scenarios and need to make sure we either cleanly get all the subsystems back to idle or everything goes into a terminal error state (depending on the fault).

**Comment by andreaskurth** [comment]
I see @cindychip has already labeled this as triaged with P2 and kept M3, but as this affects multiple blocks I'd like to ensure that we are all in agreement with resolving this post M2.5. I'm tending more towards M2.5 with P3. @GregAC @vogelpi WDYT?

**Comment by GregAC** [comment]
@andreaskurth I'm happy with post M2.5. As we've FPV here to cover these case we may not want further work here even post M2.5.

**Comment by vogelpi** [comment]
It feels out of scope for Earlgrey-PROD.M5 to do the fundamental work required for this. But it would be nice to have something better in the future. Moving to FutureReleases.
